autoprecompiles/src/bitwise_lookup_optimizer.rs

```diff
@@ -0,0 +1,139 @@
+use std::collections::BTreeSet;
+
+use itertools::Itertools;
+use powdr_ast::analyzed::AlgebraicExpression;
+use powdr_number::FieldElement;
+
+use crate::{SymbolicBusInteraction, SymbolicConstraint, SymbolicMachine, BITWISE_LOOKUP_BUS_ID};
+
+/// Optimize interactions with the bitwise lookup bus. It mostly optimizes the use of
+/// byte-range constraints.
+pub fn optimize_bitwise_lookup<T: FieldElement>(
+    mut machine: SymbolicMachine<T>,
+) -> SymbolicMachine<T> {
+    // Expressions that we need to byte-constrain at the edn.
+    let mut to_byte_constrain = vec![];
+    // Bus interaction indices we will remove.
+    let mut to_remove = vec![];
+    // New constraints (mainly substitutions) we will add.
+    let mut new_constraints: Vec<SymbolicConstraint<T>> = vec![];
+    for (index, bus_int) in machine.bus_interactions.iter().enumerate() {
+        // See: https://github.com/openvm-org/openvm/blob/v1.0.0/crates/circuits/primitives/src/bitwise_op_lookup/bus.rs
+        // Expects (x, y, z, op), where:
+        // - if op == 0, then x and y are bytes and z = 0
+        // - if op == 1, then x and y are bytes and z = x ^ y
+        if bus_int.id != BITWISE_LOOKUP_BUS_ID {
+            continue;
+        }
+        let [x, y, z, op] = &bus_int.args[..] else {
+            panic!();
+        };
+        if bus_int.mult != T::from(1).into() {
+            continue;
+        }
+
+        let AlgebraicExpression::Number(op) = op else {
+            continue;
+        };
+        // Collect all expressions to be byte-constrained and
+        // re-emit the constraints at the end in a (hopefully) more efficient way.
+        // The idea behind is that we can have duplicates or concrete numbers
+        // in those expressions.
+        if op == &0.into() {
+            // The bus interaction is equivalent to "x and y are bytes and z = 0".
+            to_byte_constrain.extend([x.clone(), y.clone()]);
+            to_remove.push(index);
+            if z != &T::from(0).into() {
```

Comment:
I think op=0 => z=0, could assert that

Reply:
No I mean this is how openvm uses it, I think it's wasteful to add a constraint for this and we should just add a Rust assertion